presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java

```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+                RowField field = prestoTypeFields.get(i);
+                fieldTypeMap.put(field.getName().get().toLowerCase(), ObjectPair.create(Integer.valueOf(i), field.getType()));
+            }
+
+            boolean outOfOrder = false;
+            ImmutableList.Builder<BlockFieldConverter> builder = ImmutableList.builder();
+            for (int i = 0, prevFieldIndex = -1; i < fieldCount; i++) {
                 parquet.schema.Type fieldType = fieldTypes.get(i);
-                converters.add(createConverter(prestoTypeParameters.get(i), columnName + "." + fieldType.getName(), fieldType, i));
+                ObjectPair<Integer, Type> pair = fieldTypeMap.get(fieldType.getName().toLowerCase());
+                if (pair == null) {
+                    log.info("Schema mismatch, metastore schema for row column %s doesn't have field %s but parquet schema has it",
+                            columnName,
+                            fieldType.getName());
+                    builder.add(ParquetDummyPrimitiveConverter.converter);
+                }
+                else {
+                    int currFieldIndex = pair.getFirst().intValue();
+                    if (prevFieldIndex > currFieldIndex) {
+                        outOfOrder = true;
+                    }
+                    builder.add(createConverter(pair.getSecond(), columnName + "." + fieldType.getName(), fieldType, currFieldIndex, true));
+                    prevFieldIndex = currFieldIndex;
+                }
             }
-            this.converters = converters.build();
+            List<BlockFieldConverter> converters = builder.build();
+            if (outOfOrder) {
+                log.info("Schema mismatch, struct field order for metastore schema for row column %s is different from that of parquet schema",
+                        columnName);
+                ImmutableList.Builder<Type> types = ImmutableList.builder();
+                converterFieldIndex = new int[fieldCount];
+                for (int i = 0, tmpIndex = 0; i < fieldCount; i++) {
+                    BlockFieldConverter converter = converters.get(i);
+                    converterFieldIndex[i] = converter.getFieldIndex();
+                    if (converterFieldIndex[i] < 0) {
+                        continue;
+                    }
+                    converter.setFieldIndex(tmpIndex++);
+                    types.add(converter.getType());
+                }
+                tmpBuilder = new InterleavedBlockBuilder(types.build(), new BlockBuilderStatus(), NULL_BUILDER_POSITIONS_THRESHOLD);
+            }
+            return converters;
+        }
+
+        public ParquetStructConverter(Type prestoType, String columnName, GroupType entryType, int fieldIndex, boolean useNames)
+        {
+            checkArgument(ROW.equals(prestoType.getTypeSignature().getBase()));
+            List<parquet.schema.Type> fieldTypes = entryType.getFields();
+
+            if (useNames) {
```

Comment:
use name based field lookup or index based field lookup
I am not sure whether we need another configuration to decide automatically support nested schema evolution (adding/removing/reordering fields in parquet), let's just use the existing useNames for now